flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java

```diff
@@ -0,0 +1,217 @@
+  // correctly commit all the data files whose checkpoint id is greater than the max committed one to iceberg table, for
+  // avoiding committing the same data files twice. This id will be attached to iceberg's meta when committing the
+  // iceberg transaction.
+  private transient long maxCommittedCheckpointId;
+
+  // A sorted map to maintain the completed data files for each pending checkpointId (which have not been committed
+  // to iceberg table). We need a sorted map here because there's possible that few checkpoints snapshot failed, for
+  // example: the 1st checkpoint have 2 data files <1, <file0, file1>>, the 2st checkpoint have 1 data files
+  // <2, <file3>>. Snapshot for checkpoint#1 interrupted because of network/disk failure etc, while we don't expect
+  // any data loss in iceberg table. So we keep the finished files <1, <file0, file1>> in memory and retry to commit
+  // iceberg table when the next checkpoint happen.
+  private final NavigableMap<Long, List<DataFile>> dataFilesPerCheckpoint = Maps.newTreeMap();
+
+  // The data files cache for current checkpoint. Once the snapshot barrier received, it will be flushed to the
+  // 'dataFilesPerCheckpoint'.
+  private final List<DataFile> dataFilesOfCurrentCheckpoint = Lists.newArrayList();
+  private transient Table table;
+
+  // All pending checkpoints states for this function.
+  private static final ListStateDescriptor<SortedMap<Long, List<DataFile>>> STATE_DESCRIPTOR = buildStateDescriptor();
+
+  private transient ListState<SortedMap<Long, List<DataFile>>> checkpointsState;
+
+  IcebergFilesCommitter(String filesCommitterUid, String fullTableName,
+                        Map<String, String> options, Configuration conf) {
+    this.filesCommitterUid = filesCommitterUid;
+    this.fullTableName = fullTableName;
+    this.options = ImmutableMap.copyOf(options);
+    this.conf = new SerializableConfiguration(conf);
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+    Catalog icebergCatalog = CATALOG_FACTORY.buildIcebergCatalog(fullTableName, options, conf.get());
+
+    table = icebergCatalog.loadTable(TableIdentifier.parse(fullTableName));
+    maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
+
+    checkpointsState = context.getOperatorStateStore().getListState(STATE_DESCRIPTOR);
+    if (context.isRestored()) {
+      maxCommittedCheckpointId = getMaxCommittedCheckpointId(table, filesCommitterUid);
+      // In the restoring path, it should have one valid snapshot for current flink job at least, so the max committed
+      // checkpoint id should be positive. If it's not positive, that means someone might have removed or expired the
+      // iceberg snapshot, in that case we should throw an exception in case of committing duplicated data files into
+      // the iceberg table.
+      Preconditions.checkArgument(maxCommittedCheckpointId > 0,
```

Comment:
IIRC, Flink checkpoint id starts at zero. So `maxCommittedCheckpointId >= 0`?

Reply:
I checked the [ZooKeeperCheckpointIDCounter](https://github.com/apache/flink/blob/39ef9593f4d88aafaa296f119d9f4a55c9731af1/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointIDCounter.java#L82) and [StandaloneCheckpointIDCounter](https://github.com/apache/flink/blob/39ef9593f4d88aafaa296f119d9f4a55c9731af1/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCheckpointIDCounter.java#L33).  Both of them should start from 1.

Reply:
Yes, but I think `maxCommittedCheckpointId != INIT_ID` should be more reasonable.